File: KidBox/KidBox/Features/Passwords/PasswordGenerator.swift
//
//  PasswordGenerator.swift
//  KidBox
//

import Foundation

/// Opzioni per la generazione (lunghezza 8–64, set di caratteri, esclusione ambigui).
struct PasswordGeneratorOptions: Equatable, Sendable {
    /// Lunghezza richiesta (viene clampata a 8…64 in `generate`).
    var length: Int = 18
    var includeUppercase: Bool = true
    var includeLowercase: Bool = true
    var includeNumbers: Bool = true
    var includeSymbols: Bool = true
    /// Esclude `0 O 1 l I` da tutti i pool (lettura più chiara, meno errori di battitura).
    var excludeAmbiguous: Bool = true

    static let `default` = PasswordGeneratorOptions()
}

enum PasswordGenerator {

    private static let lowercaseAll = "abcdefghijklmnopqrstuvwxyz"
    private static let uppercaseAll = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    private static let digitsAll = "0123456789"
    /// Simboli ASCII stampabili comuni (nessun spazio).
    private static let symbolsAll = "!@#$%^&*()-_=+[]{}|;:,.<>?/~`"

    private static let ambiguous = CharacterSet(charactersIn: "0O1lI")

    /// Genera con le opzioni di default (compatibile con il comportamento precedente, lunghezza 18).
    static func make(length: Int = 18) -> String {
        var o = PasswordGeneratorOptions.default
        o.length = length
        return generate(options: o)
    }

    /// Costruisce i pool in base alle opzioni; se nessun set è attivo, usa alfanumerico + simboli.
    static func generate(options raw: PasswordGeneratorOptions) -> String {
        var options = raw
        options.length = max(8, min(64, options.length))

        var pools: [String] = []
        if options.includeLowercase {
            pools.append(filtered(lowercaseAll, excludeAmbiguous: options.excludeAmbiguous))
        }
        if options.includeUppercase {
            pools.append(filtered(uppercaseAll, excludeAmbiguous: options.excludeAmbiguous))
        }
        if options.includeNumbers {
            pools.append(filtered(digitsAll, excludeAmbiguous: options.excludeAmbiguous))
        }
        if options.includeSymbols {
            pools.append(symbolsAll)
        }

        if pools.isEmpty {
            var fallback = PasswordGeneratorOptions.default
            fallback.length = options.length
            return generate(options: fallback)
        }

        // Rimuovi pool vuoti (es. tutto filtrato via).
        pools = pools.filter { !$0.isEmpty }
        if pools.isEmpty {
            return make(length: options.length)
        }

        var rng = SystemRandomNumberGenerator()
        var chars: [Character] = []
        chars.reserveCapacity(options.length)

        // Garantisce almeno un carattere da ogni pool abilitato e non vuoto.
        for p in pools {
            guard let ch = randomChar(from: p, using: &rng) else { continue }
            chars.append(ch)
        }

        let union = pools.joined()
        guard !union.isEmpty else { return make(length: options.length) }

        while chars.count < options.length {
            guard let ch = randomChar(from: union, using: &rng) else { break }
            chars.append(ch)
        }

        chars.shuffle(using: &rng)
        return String(chars.prefix(options.length))
    }

    private static func filtered(_ s: String, excludeAmbiguous: Bool) -> String {
        guard excludeAmbiguous else { return s }
        return String(s.unicodeScalars.filter { !ambiguous.contains($0) })
    }

    private static func randomChar(from s: String, using rng: inout SystemRandomNumberGenerator) -> Character? {
        guard !s.isEmpty else { return nil }
        let idx = Int.random(in: 0..<s.count, using: &rng)
        let i = s.index(s.startIndex, offsetBy: idx)
        return s[i]
    }
}

File: KidboxWebApp/src/passwordGenerator.js
/**
 * Generatore di password, porting di `PasswordGenerator.swift`: stessi pool di
 * caratteri, stessa esclusione degli ambigui, stessa garanzia di almeno un
 * carattere per ogni set attivo.
 *
 * L'unica differenza è la sorgente casuale: qui `crypto.getRandomValues` al
 * posto di `SystemRandomNumberGenerator`.
 */
const LOWERCASE = "abcdefghijklmnopqrstuvwxyz";
const UPPERCASE = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const DIGITS = "0123456789";
const SYMBOLS = "!@#$%^&*()-_=+[]{}|;:,.<>?/~`";
const AMBIGUOUS = new Set(["0", "O", "1", "l", "I"]);

export const DEFAULT_OPTIONS = {
  length: 18,
  includeUppercase: true,
  includeLowercase: true,
  includeNumbers: true,
  includeSymbols: true,
  excludeAmbiguous: true,
};

/** Indice casuale senza modulo bias. */
function randomIndex(max) {
  if (max <= 0) return 0;
  const limit = Math.floor(0xffffffff / max) * max;
  const buf = new Uint32Array(1);
  let value;
  do {
    crypto.getRandomValues(buf);
    [value] = buf;
  } while (value >= limit);
  return value % max;
}

function pick(pool) {
  return pool[randomIndex(pool.length)];
}

function filtered(pool, excludeAmbiguous) {
  if (!excludeAmbiguous) return pool;
  return [...pool].filter((c) => !AMBIGUOUS.has(c)).join("");
}

export function generate(raw = DEFAULT_OPTIONS) {
  const options = { ...DEFAULT_OPTIONS, ...raw };
  options.length = Math.max(8, Math.min(64, options.length || 18));

  let pools = [];
  if (options.includeLowercase) pools.push(filtered(LOWERCASE, options.excludeAmbiguous));
  if (options.includeUppercase) pools.push(filtered(UPPERCASE, options.excludeAmbiguous));
  if (options.includeNumbers) pools.push(filtered(DIGITS, options.excludeAmbiguous));
  // I simboli non contengono ambigui: nessun filtro, come su iOS.
  if (options.includeSymbols) pools.push(SYMBOLS);
  pools = pools.filter((p) => p.length > 0);

  // Nessun set attivo: si ricade sui default invece di restituire vuoto.
  if (pools.length === 0) {
    return generate({ ...DEFAULT_OPTIONS, length: options.length });
  }

  // Almeno un carattere da ogni set abilitato, poi si riempie dall'unione.
  const chars = pools.map(pick);
  const union = pools.join("");
  while (chars.length < options.length) chars.push(pick(union));

  for (let i = chars.length - 1; i > 0; i -= 1) {
    const j = randomIndex(i + 1);
    [chars[i], chars[j]] = [chars[j], chars[i]];
  }
  return chars.slice(0, options.length).join("");
}
